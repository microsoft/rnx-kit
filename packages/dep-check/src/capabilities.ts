import type { Capability, KitCapabilities } from "@rnx-kit/config";
import type { PackageManifest } from "@rnx-kit/tools-node/package";
import semver from "semver";
import { getProfilesFor, getProfileVersionsFor } from "./profiles";
import { concatVersionRanges, keysOf } from "./helpers";
import type { CapabilitiesOptions, Package, Profile } from "./types";

export function capabilitiesFor(
  { dependencies, devDependencies, peerDependencies }: PackageManifest,
  { kitType = "library", customProfilesPath }: CapabilitiesOptions = {}
): Partial<KitCapabilities> | undefined {
  const targetReactNativeVersion =
    peerDependencies?.["react-native"] || dependencies?.["react-native"];
  if (!targetReactNativeVersion) {
    return undefined;
  }

  const profiles = getProfilesFor(targetReactNativeVersion, customProfilesPath);
  const packageToCapabilityMap: Record<string, Capability[]> = {};
  profiles.forEach((profile) => {
    keysOf(profile).reduce((result, capability) => {
      const { name } = profile[capability];
      if (!result[name]) {
        result[name] = [capability];
      } else {
        result[name].push(capability);
      }
      return result;
    }, packageToCapabilityMap);
  });

  const reactNativeVersion = concatVersionRanges(
    getProfileVersionsFor(targetReactNativeVersion)
  );

  return {
    reactNativeVersion,
    ...(kitType === "library"
      ? {
          reactNativeDevVersion:
            devDependencies?.["react-native"] ||
            semver.minVersion(reactNativeVersion)?.version,
        }
      : undefined),
    kitType,
    capabilities: Array.from(
      keysOf({
        ...dependencies,
        ...peerDependencies,
        ...devDependencies,
      }).reduce<Set<Capability>>((result, dependency) => {
        if (dependency in packageToCapabilityMap) {
          packageToCapabilityMap[dependency].forEach((capability) => {
            result.add(capability);
          });
        }
        return result;
      }, new Set<Capability>())
    ).sort(),
  };
}

export function resolveCapabilities(
  capabilities: Capability[],
  profiles: Profile[]
): Record<string, Package[]> {
  const unresolvedCapabilities = new Set<string>();
  const packages = capabilities.reduce<Record<string, Package[]>>(
    (dependencies, capability) => {
      profiles.forEach((profile) => {
        const pkg = profile[capability];
        if (!pkg) {
          unresolvedCapabilities.add(capability);
          return;
        }

        const { name, version } = pkg;
        if (name in dependencies) {
          const versions = dependencies[name];
          if (!versions.find((current) => current.version === version)) {
            versions.push(pkg);
          }
        } else {
          dependencies[name] = [pkg];
        }
      });
      return dependencies;
    },
    {}
  );

  if (unresolvedCapabilities.size > 0) {
    const message = Array.from(unresolvedCapabilities).reduce(
      (lines, capability) => (lines += `\n    ${capability}`),
      "The following capabilities could not be resolved for one or more profiles:"
    );

    console.warn(message);
  }

  return packages;
}
