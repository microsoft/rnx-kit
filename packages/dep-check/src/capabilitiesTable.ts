import type { Capability } from "@rnx-kit/config";
import { ProfileVersion, getAllProfiles } from "./profiles";

function maxStringLength(currentMax: number, str: string): number {
  return Math.max(currentMax, str.length);
}

function sortCoreFirst(lhs: string, rhs: string): number {
  if (lhs.startsWith("core-")) {
    if (!rhs.startsWith("core-")) {
      return -1;
    }
  } else if (rhs.startsWith("core-")) {
    return 1;
  }

  if (lhs === rhs) {
    return 0;
  }

  return lhs < rhs ? -1 : 1;
}

const allProfiles = getAllProfiles();
const allVersions = Object.keys(allProfiles).reverse() as ProfileVersion[];
const allCapabilities = Object.keys(allProfiles["0.64"]).sort(
  sortCoreFirst
) as Capability[];

const headers = ["Capability", ...allVersions];

const columns = allVersions.reduce<string[][]>(
  (result, version) => {
    const profile = allProfiles[version];
    const column = allCapabilities.map((capability) => {
      const { name, version } = profile[capability];
      return `${name}@${version}`;
    });
    result.push(column);
    return result;
  },
  [allCapabilities]
);

const columnWidths = columns.map((column, index) =>
  column.reduce(maxStringLength, headers[index].length)
);

const pad = (s: string, index: number) => s.padEnd(columnWidths[index]);

const table = [
  `| ${headers.map(pad).join(" | ")} |`,
  `| ${columnWidths.map((width) => "-".repeat(width)).join(" | ")} |`,
  ...allCapabilities.map(
    (_, capability) =>
      `| ${columns
        .map((column, index) => pad(column[capability], index))
        .join(" | ")} |`
  ),
].join("\n");

console.log(table);
