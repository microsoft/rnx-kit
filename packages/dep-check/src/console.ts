import chalk from "chalk";

export function error(message: string): void {
  console.error(chalk.red("error:"), message);
}

export function warn(message: string): void {
  console.warn(chalk.yellow("warn:"), message);
}
