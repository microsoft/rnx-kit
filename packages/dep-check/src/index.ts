export { capabilitiesFor } from "./capabilities";
export { checkPackageManifest } from "./check";
export { Args, cli } from "./cli";
export { updatePackageManifest } from "./manifest";
export * from "./types";
