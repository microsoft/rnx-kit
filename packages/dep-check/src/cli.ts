#!/usr/bin/env node

import type { KitType } from "@rnx-kit/config";
import { error, warn } from "@rnx-kit/console";
import { hasProperty } from "@rnx-kit/tools-language/properties";
import { findPackageDir } from "@rnx-kit/tools-node/package";
import {
  findWorkspacePackages,
  findWorkspaceRoot,
} from "@rnx-kit/tools-workspaces";
import isString from "lodash/isString";
import * as path from "path";
import { makeCheckCommand } from "./check";
import { initializeConfig } from "./initialize";
import { resolveCustomProfiles } from "./profiles";
import { makeSetVersionCommand } from "./setVersion";
import type { Args, Command } from "./types";
import { makeVigilantCommand } from "./vigilant";

function ensureKitType(type: string): KitType | undefined {
  switch (type) {
    case "app":
    case "library":
      return type;
    default:
      return undefined;
  }
}

async function getManifests(
  packages: (string | number)[] | undefined
): Promise<string[] | undefined> {
  if (Array.isArray(packages)) {
    return packages.reduce<string[]>((result, pkg) => {
      const dir = findPackageDir(pkg.toString());
      if (dir) {
        result.push(path.join(dir, "package.json"));
      }
      return result;
    }, []);
  }

  const packageDir = findPackageDir();
  if (!packageDir) {
    return undefined;
  }

  // Make sure we don't return all packages when dep-check is run inside a
  // package that just happened to be part of a workspace.
  const currentPackageJson = path.join(packageDir, "package.json");
  try {
    if ((await findWorkspaceRoot()) !== packageDir) {
      return [currentPackageJson];
    }
  } catch (_) {
    return [currentPackageJson];
  }

  try {
    const allPackages = (await findWorkspacePackages()).map((p) =>
      path.join(p, "package.json")
    );
    allPackages.push(currentPackageJson);
    return allPackages;
  } catch (e) {
    if (hasProperty(e, "message")) {
      error(e.message);
      return undefined;
    }

    throw e;
  }
}

function makeInitializeCommand(
  kitType: string,
  customProfiles: string | undefined
): Command | undefined {
  const verifiedKitType = ensureKitType(kitType);
  if (!verifiedKitType) {
    error(`Invalid kit type: '${kitType}'`);
    return undefined;
  }

  return (manifest: string) => {
    initializeConfig(manifest, {
      kitType: verifiedKitType,
      customProfilesPath: customProfiles,
    });
    return 0;
  };
}

function reportConflicts(conflicts: [string, string][], args: Args): boolean {
  return conflicts.reduce<boolean>((result, [lhs, rhs]) => {
    if (lhs in args && rhs in args) {
      error(`--${lhs} and --${rhs} cannot both be specified at the same time.`);
      return true;
    }
    return result;
  }, false);
}

async function makeCommand(args: Args): Promise<Command | undefined> {
  const conflicts: [string, string][] = [
    ["init", "vigilant"],
    ["init", args.write ? "write" : "no-write"],
    ["set-version", args.write ? "write" : "no-write"],
  ];
  if (reportConflicts(conflicts, args)) {
    return undefined;
  }

  const {
    "custom-profiles": customProfiles,
    "exclude-packages": excludePackages,
    init,
    loose,
    "set-version": setVersion,
    vigilant,
    write,
  } = args;

  if (isString(init)) {
    return makeInitializeCommand(init, customProfiles?.toString());
  }

  // When `--set-version` is without a value, `setVersion` is an empty string if
  // invoked directly. When invoked via `@react-native-community/cli`,
  // `setVersion` is `true` instead.
  if (setVersion || isString(setVersion)) {
    return makeSetVersionCommand(setVersion);
  }

  if (isString(vigilant)) {
    const customProfilesPath = resolveCustomProfiles(
      process.cwd(),
      customProfiles?.toString()
    );
    return makeVigilantCommand({
      customProfiles: customProfilesPath,
      excludePackages: excludePackages?.toString(),
      loose,
      versions: vigilant.toString(),
      write,
    });
  }

  return makeCheckCommand({ loose, write });
}

export async function cli({ packages, ...args }: Args): Promise<void> {
  const command = await makeCommand(args);
  if (!command) {
    process.exit(1);
  }

  const manifests = await getManifests(packages);
  if (!manifests) {
    error("Could not find package root");
    process.exit(1);
  }

  // We will optimistically run through all packages regardless of failures. In
  // most scenarios, this should be fine: Both init and check+write write to
  // disk only when everything is in order for the target package. Packages with
  // invalid or missing configurations are skipped.
  const exitCode = manifests.reduce((exitCode: number, manifest: string) => {
    try {
      return command(manifest) || exitCode;
    } catch (e) {
      if (hasProperty(e, "message")) {
        const currentPackageJson = path.relative(process.cwd(), manifest);

        if (hasProperty(e, "code") && e.code === "ENOENT") {
          warn(`${currentPackageJson}: ${e.message}`);
          return exitCode;
        }

        error(`${currentPackageJson}: ${e.message}`);
        return exitCode || 1;
      }

      throw e;
    }
  }, 0);

  process.exit(exitCode);
}

if (require.main === module) {
  require("yargs").usage(
    "$0 [packages...]",
    "Dependency checker for React Native apps",
    {
      "custom-profiles": {
        description:
          "Path to custom profiles. This can be a path to a JSON file, a `.js` file, or a module name.",
        type: "string",
        requiresArg: true,
      },
      "exclude-packages": {
        description:
          "Comma-separated list of package names to exclude from inspection.",
        type: "string",
        requiresArg: true,
        implies: "vigilant",
      },
      init: {
        description:
          "Writes an initial kit config to the specified 'package.json'.",
        choices: ["app", "library"],
        conflicts: ["vigilant"],
      },
      loose: {
        default: false,
        description:
          "Determines how strict the React Native version requirement should be. Useful for apps that depend on a newer React Native version than their dependencies declare support for.",
        type: "boolean",
      },
      "set-version": {
        description:
          "Sets `reactNativeVersion` and `reactNativeDevVersion` for any configured package. There is an interactive prompt if no value is provided. The value should be a comma-separated list of `react-native` versions to set, where the first number specifies the development version. Example: `0.64,0.63`",
        type: "string",
        conflicts: ["init", "vigilant"],
      },
      vigilant: {
        description:
          "Inspects packages regardless of whether they've been configured. Specify a comma-separated list of profile versions to compare against, e.g. `0.63,0.64`. The first number specifies the target version.",
        type: "string",
        requiresArg: true,
        conflicts: ["init"],
      },
      write: {
        default: false,
        description: "Writes changes to the specified 'package.json'.",
        type: "boolean",
      },
    },
    cli
  ).argv;
}
