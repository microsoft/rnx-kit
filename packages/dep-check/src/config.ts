import type { KitConfig } from "@rnx-kit/config";
import semver from "semver";
import { getRequirements } from "./dependencies";
import type { PackageManifest } from "./types";

type KitManifest = Required<
  Pick<
    KitConfig,
    "capabilities" | "kitType" | "reactNativeVersion" | "reactNativeDevVersion"
  >
>;

function minVersion(version: string): string {
  const v = semver.minVersion(version)?.version;
  return v || "0.0.0";
}

export function parseConfig(
  {
    capabilities = [],
    kitType = "library",
    reactNativeVersion = "",
    reactNativeDevVersion,
  }: KitConfig,
  manifest: PackageManifest,
  projectRoot: string
): KitManifest {
  switch (kitType) {
    case "app": {
      const {
        reactNativeVersion,
        capabilities: requiredCapabilities,
      } = getRequirements(manifest, projectRoot);
      requiredCapabilities.push(...capabilities);
      return {
        capabilities: requiredCapabilities,
        kitType,
        reactNativeVersion,
        reactNativeDevVersion:
          reactNativeDevVersion || minVersion(reactNativeVersion),
      };
    }

    case "library": {
      return {
        capabilities,
        kitType,
        reactNativeVersion,
        reactNativeDevVersion:
          reactNativeDevVersion || minVersion(reactNativeVersion),
      };
    }
  }
}
