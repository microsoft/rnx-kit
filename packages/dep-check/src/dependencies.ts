import type { Capability, KitConfig, KitType } from "@rnx-kit/config";
import { getKitCapabilities, getKitConfig } from "@rnx-kit/config";
import { error, warn } from "@rnx-kit/console";
import type { PackageManifest } from "@rnx-kit/tools-node/package";
import {
  findPackageDependencyDir,
  readPackage,
} from "@rnx-kit/tools-node/package";
import { concatVersionRanges } from "./helpers";
import {
  getProfilesFor,
  getProfileVersionsFor,
  profilesSatisfying,
} from "./profiles";
import type { CheckOptions, Profile, ProfileVersion } from "./types";

type Requirements = Required<
  Pick<KitConfig, "reactNativeVersion" | "capabilities">
>;

type Trace = {
  module: string;
  reactNativeVersion: string;
  profiles: ProfileVersion[];
};

function isCoreCapability(capability: Capability): boolean {
  return capability.startsWith("core-");
}

function isDevOnlyCapability(
  capability: Capability,
  profiles: Partial<Profile>[]
): boolean {
  return profiles.some((profile) => profile[capability]?.devOnly);
}

export function visitDependencies(
  { dependencies }: PackageManifest,
  projectRoot: string,
  visitor: (module: string, modulePath: string) => void,
  visited: Set<string> = new Set<string>()
): void {
  if (!dependencies) {
    return;
  }

  Object.keys(dependencies).forEach((dependency) => {
    if (visited.has(dependency)) {
      return;
    }

    visited.add(dependency);

    const packageDir = findPackageDependencyDir(dependency, {
      startDir: projectRoot,
      resolveSymlinks: true,
    });
    if (!packageDir) {
      warn(`Unable to resolve module '${dependency}'`);
      return;
    }

    visitor(dependency, packageDir);

    const manifest = readPackage(packageDir);
    visitDependencies(manifest, packageDir, visitor, visited);
  });
}

export function getRequirements(
  targetReactNativeVersion: string,
  kitType: KitType,
  targetManifest: PackageManifest,
  projectRoot: string,
  customProfiles: string | undefined,
  { loose }: Pick<CheckOptions, "loose">
): Requirements {
  let profileVersions = getProfileVersionsFor(targetReactNativeVersion);
  if (profileVersions.length === 0) {
    throw new Error(
      `No profile could satisfy React Native version: ${targetReactNativeVersion}`
    );
  }

  const allCapabilities = new Set<Capability>();
  if (kitType === "app") {
    const trace: Trace[] = [
      {
        module: targetManifest.name,
        reactNativeVersion: targetReactNativeVersion,
        profiles: profileVersions,
      },
    ];

    visitDependencies(targetManifest, projectRoot, (module, modulePath) => {
      const kitConfig = getKitConfig({ cwd: modulePath });
      if (!kitConfig?.reactNativeVersion) {
        return;
      }

      const { reactNativeVersion, capabilities } =
        getKitCapabilities(kitConfig);

      const validVersions = profilesSatisfying(
        profileVersions,
        reactNativeVersion
      );
      if (validVersions.length != profileVersions.length) {
        trace.push({
          module,
          reactNativeVersion,
          profiles: validVersions,
        });
      }

      if (Array.isArray(capabilities)) {
        capabilities.forEach((capability) => allCapabilities.add(capability));
      }

      // In strict mode, we want to continue so we can catch all dependencies
      // that cannot be satisfied. Whereas in loose mode, we can ignore them and
      // carry on with the profiles that satisfy the rest.
      if (validVersions.length > 0) {
        profileVersions = validVersions;
      }
    });

    if (trace[trace.length - 1].profiles.length === 0) {
      const message = "No React Native profile could satisfy all dependencies";
      const fullTrace = [
        message,
        ...trace.map(({ module, reactNativeVersion, profiles }) => {
          const satisfiedVersions = profiles.join(", ");
          return `    [${satisfiedVersions}] satisfies '${module}' because it supports '${reactNativeVersion}'`;
        }),
      ].join("\n");
      if (loose) {
        warn(fullTrace);
      } else {
        error(fullTrace);
        throw new Error(message);
      }
    }

    const profiles = getProfilesFor(profileVersions, customProfiles);
    allCapabilities.forEach((capability) => {
      /**
       * Core capabilities are capabilities that must always be declared by the
       * hosting app and should not be included when gathering requirements.
       * This is to avoid forcing an app to install dependencies it does not
       * need, e.g. `react-native-windows` when the app only supports iOS.
       */
      if (
        isCoreCapability(capability) ||
        isDevOnlyCapability(capability, profiles)
      ) {
        allCapabilities.delete(capability);
      }
    });
  }

  return {
    reactNativeVersion: concatVersionRanges(profileVersions),
    capabilities: Array.from(allCapabilities),
  };
}
