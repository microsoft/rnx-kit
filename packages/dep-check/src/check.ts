import { getKitCapabilities, getKitConfig } from "@rnx-kit/config";
import { error, info, warn } from "@rnx-kit/console";
import { isPackageManifest, readPackage } from "@rnx-kit/tools-node/package";
import chalk from "chalk";
import { diffLinesUnified } from "jest-diff";
import path from "path";
import { getRequirements } from "./dependencies";
import { findBadPackages } from "./findBadPackages";
import { modifyManifest, printMigrationMessage } from "./helpers";
import { updatePackageManifest } from "./manifest";
import { getProfilesFor, resolveCustomProfiles } from "./profiles";
import type { CheckConfig, CheckOptions, Command } from "./types";

export function getCheckConfig(
  manifestPath: string,
  {
    loose,
    uncheckedReturnCode = 0,
    supportedVersions,
    targetVersion,
  }: CheckOptions
): number | CheckConfig {
  const manifest = readPackage(manifestPath);
  if (!isPackageManifest(manifest)) {
    error(
      `'${manifestPath}' does not contain a valid package manifest - please make sure it's not missing 'name' or 'version'`
    );
    return 1;
  }

  const badPackages = findBadPackages(manifest);
  if (badPackages) {
    warn(
      `Known bad packages are found in '${manifest.name}':\n` +
        badPackages
          .map((pkg) => `    ${pkg.name}@${pkg.version}: ${pkg.reason}`)
          .join("\n")
    );
  }

  const projectRoot = path.dirname(manifestPath);
  const kitConfig = getKitConfig({ cwd: projectRoot });
  if (!kitConfig) {
    return uncheckedReturnCode;
  }

  const {
    reactNativeVersion: targetReactNativeVersion,
    reactNativeDevVersion,
    kitType,
    capabilities: targetCapabilities,
    customProfiles,
  } = getKitCapabilities({
    // React Native versions declared in the package's config should always
    // override the ones specified with the `--vigilant` flag.
    ...(supportedVersions
      ? { reactNativeVersion: supportedVersions }
      : undefined),
    // We should not set dev version if the package is configured. It may have
    // been intentionally left out to reuse `reactNativeVersion`.
    ...(!kitConfig.reactNativeVersion && targetVersion
      ? { reactNativeDevVersion: targetVersion }
      : undefined),
    ...kitConfig,
  });

  const customProfilesPath = resolveCustomProfiles(projectRoot, customProfiles);

  const { reactNativeVersion, capabilities: requiredCapabilities } =
    getRequirements(
      targetReactNativeVersion,
      kitType,
      manifest,
      projectRoot,
      customProfilesPath,
      { loose }
    );
  requiredCapabilities.push(...targetCapabilities);

  return {
    kitType,
    reactNativeVersion,
    reactNativeDevVersion,
    capabilities: requiredCapabilities,
    customProfilesPath,
    manifest,
  };
}

export function checkPackageManifest(
  manifestPath: string,
  options: CheckOptions
): number {
  const result = options.config || getCheckConfig(manifestPath, options);
  if (typeof result === "number") {
    return result;
  }

  const {
    kitType,
    reactNativeVersion,
    reactNativeDevVersion,
    capabilities,
    customProfilesPath,
    manifest,
  } = result;

  if (capabilities.length === 0) {
    return options.uncheckedReturnCode || 0;
  }

  const updatedManifest = updatePackageManifest(
    manifest,
    capabilities,
    getProfilesFor(reactNativeVersion, customProfilesPath),
    getProfilesFor(reactNativeDevVersion, customProfilesPath),
    kitType
  );

  // Don't fail when manifests only have whitespace differences.
  const updatedManifestJson = JSON.stringify(updatedManifest, undefined, 2);
  const normalizedManifestJson = JSON.stringify(manifest, undefined, 2);

  if (updatedManifestJson !== normalizedManifestJson) {
    if (options.write) {
      modifyManifest(manifestPath, updatedManifest);
    } else {
      const diff = diffLinesUnified(
        normalizedManifestJson.split("\n"),
        updatedManifestJson.split("\n"),
        {
          aAnnotation: "Current",
          aColor: chalk.red,
          bAnnotation: "Expected",
          bColor: chalk.green,
        }
      );
      console.log(diff);

      error(
        "Changes are needed to satisfy all requirements. Re-run with `--write` to have dep-check apply them."
      );

      const url = chalk.bold("https://aka.ms/dep-check");
      info(`Visit ${url} for more information about dep-check.`);

      printMigrationMessage();

      return 1;
    }
  }

  printMigrationMessage();

  return 0;
}

export function makeCheckCommand(options: CheckOptions): Command {
  return (manifest: string) => {
    return checkPackageManifest(manifest, options);
  };
}
