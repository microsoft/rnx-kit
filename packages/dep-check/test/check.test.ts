import { checkPackageManifest } from "../src/check";
import profile_0_62 from "../src/profiles/profile-0.62";
import profile_0_63 from "../src/profiles/profile-0.63";
import profile_0_64 from "../src/profiles/profile-0.64";

jest.mock("fs");

describe("checkPackageManifest({ kitType: 'library' })", () => {
  const rnxKitConfig = require("@rnx-kit/config");
  const fs = require("fs");

  const consoleErrorSpy = jest.spyOn(global.console, "error");
  const consoleLogSpy = jest.spyOn(global.console, "log");
  const consoleWarnSpy = jest.spyOn(global.console, "warn");

  const defaultOptions = { loose: false, write: false };

  const mockManifest = {
    name: "@rnx-kit/dep-check",
    version: "0.0.1",
  };

  const v62_v63_v64 = [
    profile_0_62["core-ios"].version,
    profile_0_63["core-ios"].version,
    profile_0_64["core-ios"].version,
  ].join(" || ");

  beforeEach(() => {
    consoleErrorSpy.mockReset();
    consoleLogSpy.mockReset();
    consoleWarnSpy.mockReset();
    fs.__setMockContent({});
    rnxKitConfig.__setMockConfig();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test("returns error code when reading invalid manifests", () => {
    expect(checkPackageManifest("package.json", defaultOptions)).not.toBe(0);
    expect(consoleErrorSpy).toBeCalledTimes(1);
  });

  test("returns early if 'rnx-kit' is missing from the manifest", () => {
    fs.__setMockContent({
      ...mockManifest,
      dependencies: { "react-native-linear-gradient": "0.0.0" },
    });

    const options = { ...defaultOptions, uncheckedReturnCode: -1 };
    expect(checkPackageManifest("package.json", options)).toBe(-1);
    expect(consoleWarnSpy).toBeCalled();
  });

  test("prints warnings when detecting bad packages", () => {
    fs.__setMockContent({
      ...mockManifest,
      dependencies: { "react-native-linear-gradient": "0.0.0" },
      peerDependencies: {
        "react-native": profile_0_64["core-ios"],
      },
      devDependencies: {
        "react-native": profile_0_64["core-ios"],
      },
    });
    rnxKitConfig.__setMockConfig({ reactNativeVersion: "0.64.0" });

    expect(checkPackageManifest("package.json", defaultOptions)).toBe(0);
    expect(consoleErrorSpy).not.toBeCalled();
    expect(consoleWarnSpy.mock.calls).toMatchSnapshot();
  });

  test("prints warnings when detecting bad packages (with version range)", () => {
    fs.__setMockContent({
      ...mockManifest,
      dependencies: { "react-native-linear-gradient": "0.0.0" },
    });
    rnxKitConfig.__setMockConfig({ reactNativeVersion: "^0.63.0 || ^0.64.0" });

    expect(checkPackageManifest("package.json", defaultOptions)).toBe(0);
    expect(consoleErrorSpy).not.toBeCalled();
    expect(consoleWarnSpy.mock.calls).toMatchSnapshot();
  });

  test("returns early if no capabilities are defined", () => {
    fs.__setMockContent(mockManifest);
    rnxKitConfig.__setMockConfig({ reactNativeVersion: "0.64.0" });

    expect(checkPackageManifest("package.json", defaultOptions)).toBe(0);
    expect(consoleErrorSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("returns if no changes are needed", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        "react-native": profile_0_64["core-ios"].version,
      },
      devDependencies: {
        "react-native": profile_0_64["core-ios"].version,
      },
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "0.64.0",
      capabilities: ["core-ios"],
    });

    expect(checkPackageManifest("package.json", defaultOptions)).toBe(0);
    expect(consoleErrorSpy).not.toBeCalled();
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("returns if no changes are needed (write: true)", () => {
    let didWriteToPath = false;

    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        "react-native": profile_0_64["core-ios"].version,
      },
      devDependencies: {
        "react-native": profile_0_64["core-ios"].version,
      },
    });
    fs.__setMockFileWriter((p, _content) => {
      didWriteToPath = p;
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "0.64.0",
      capabilities: ["core-ios"],
    });

    expect(
      checkPackageManifest("package.json", { loose: false, write: true })
    ).toBe(0);
    expect(didWriteToPath).toBe(false);
    expect(consoleErrorSpy).not.toBeCalled();
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("returns error code if changes are needed", () => {
    fs.__setMockContent(mockManifest);
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "0.64.0",
      capabilities: ["core-ios"],
    });

    expect(checkPackageManifest("package.json", defaultOptions)).not.toBe(0);
    expect(consoleErrorSpy).toBeCalledTimes(1);
    expect(consoleWarnSpy).not.toBeCalled();
    expect(consoleLogSpy).toBeCalledTimes(2);
  });

  test("writes changes back to 'package.json'", () => {
    let didWriteToPath = false;

    fs.__setMockContent(mockManifest);
    fs.__setMockFileWriter((p, _content) => {
      didWriteToPath = p;
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "0.64.0",
      capabilities: ["core-ios"],
    });

    expect(
      checkPackageManifest("package.json", { loose: false, write: true })
    ).toBe(0);
    expect(didWriteToPath).toBe("package.json");
    expect(consoleErrorSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
    expect(consoleLogSpy).not.toBeCalled();
  });

  test("uses minimum supported version as development version", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        "react-native": v62_v63_v64,
      },
      devDependencies: {
        "react-native": profile_0_62["core-ios"].version,
      },
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "^0.62 || ^0.63 || ^0.64",
      capabilities: ["core-ios"],
    });

    expect(checkPackageManifest("package.json", defaultOptions)).toBe(0);
    expect(consoleErrorSpy).not.toBeCalled();
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("uses declared development version", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        "react-native": v62_v63_v64,
      },
      devDependencies: {
        "react-native": profile_0_63["core-ios"].version,
      },
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "^0.62 || ^0.63 || ^0.64",
      reactNativeDevVersion: "0.63.4",
      capabilities: ["core-ios"],
    });

    expect(checkPackageManifest("package.json", defaultOptions)).toBe(0);
    expect(consoleErrorSpy).not.toBeCalled();
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("handles development version ranges", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        "react-native": v62_v63_v64,
      },
      devDependencies: {
        "react-native": profile_0_63["core-ios"].version,
      },
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "^0.62 || ^0.63 || ^0.64",
      reactNativeDevVersion: "^0.63.4",
      capabilities: ["core-ios"],
    });

    expect(checkPackageManifest("package.json", defaultOptions)).toBe(0);
    expect(consoleErrorSpy).not.toBeCalled();
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });
});
