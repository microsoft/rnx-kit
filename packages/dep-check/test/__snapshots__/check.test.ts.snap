// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkPackageManifest({ kitType: 'library' }) preserves indentation in 'package.json' 1`] = `
"{
	\\"name\\": \\"@rnx-kit/dep-check\\",
	\\"version\\": \\"0.0.1\\",
	\\"peerDependencies\\": {
		\\"react-native\\": \\"^0.64.2\\"
	},
	\\"devDependencies\\": {
		\\"react-native\\": \\"^0.64.2\\"
	}
}
"
`;

exports[`checkPackageManifest({ kitType: 'library' }) prints warnings when detecting bad packages (with version range) 1`] = `
Array [
  Array [
    "warn",
    "Known bad packages are found in '@rnx-kit/dep-check':
    react-native-linear-gradient@*: This package is unmaintained and causes significant degradation in app start up time.",
  ],
]
`;

exports[`checkPackageManifest({ kitType: 'library' }) prints warnings when detecting bad packages 1`] = `
Array [
  Array [
    "warn",
    "Known bad packages are found in '@rnx-kit/dep-check':
    react-native-linear-gradient@*: This package is unmaintained and causes significant degradation in app start up time.",
  ],
]
`;
